File: src/commit-reveal.ts
// Commit-reveal protocol for sequencer-proof MEV resistance.
//
// Phase 1 (block N):       trader submits hash(side ‖ size ‖ limit ‖ nonce ‖ trader)
// Phase 2 (block ≤ N + K): trader reveals all parts; matcher checks hash
// Phase 3 (next batch):    revealed order enters the FBA buffer
//
// Sequencer cannot front-run because the hash hides every value. To prevent
// commit spam (commit-without-reveal grief), each commit posts a small bond
// that is seized if no valid reveal lands within K batches.
//
// Censorship resistance: a reveal can also be force-included on L1; on the
// next ER session sync the matcher honors it with the original commit timestamp
// for ordering purposes.

import { commitHash } from './math.ts';
import type { CommitEntry, Order, Side } from './types.ts';

export interface RevealPayload {
  readonly market: string;
  readonly trader: string;
  readonly side: Side;
  readonly size: number;
  readonly limitPrice: number;
  readonly nonce: string;
}

export function buildCommitHash(payload: RevealPayload): string {
  return commitHash([
    payload.market,
    payload.trader,
    payload.side,
    payload.size.toString(),
    payload.limitPrice.toString(),
    payload.nonce,
  ]);
}

export class CommitRevealRegistry {
  private readonly commits = new Map<string, CommitEntry>();
  private readonly seizedBonds: Array<{ hash: string; trader: string; bond: number; batchNum: number }> = [];

  registerCommit(args: {
    hash: string;
    trader: string;
    market: string;
    bondLamports: number;
    currentBatch: number;
    expireInBatches: number;
  }): void {
    if (this.commits.has(args.hash)) {
      throw new Error(`Commit already exists for hash ${args.hash}`);
    }
    this.commits.set(args.hash, {
      hash: args.hash,
      trader: args.trader,
      market: args.market,
      bondLamports: args.bondLamports,
      committedAtBatch: args.currentBatch,
      expireAtBatch: args.currentBatch + args.expireInBatches,
    });
  }

  /**
   * Verify a reveal against the prior commit. If valid, return a synthesized
   * taker order to enqueue into the next batch and remove the commit.
   *
   * Returns null if the reveal does not match a committed hash.
   */
  redeem(args: {
    payload: RevealPayload;
    currentBatch: number;
    nowMs: number;
    orderIdSeq: number;
  }): Order | null {
    const expectedHash = buildCommitHash(args.payload);
    const entry = this.commits.get(expectedHash);
    if (!entry) return null;
    if (entry.trader !== args.payload.trader) return null;
    if (entry.market !== args.payload.market) return null;
    if (args.currentBatch > entry.expireAtBatch) {
      // Reveal arrived too late; commit will be forfeited on next sweep.
      return null;
    }
    this.commits.delete(expectedHash);
    return {
      id: `reveal_${args.orderIdSeq}_b${args.currentBatch}`,
      market: args.payload.market,
      trader: args.payload.trader,
      side: args.payload.side,
      size: args.payload.size,
      limitPrice: args.payload.limitPrice,
      type: 'taker',
      timestamp: args.nowMs,
      postOnly: false,
    };
  }

  /** Sweep expired commits and seize bonds. */
  sweepExpired(currentBatch: number): Array<{ hash: string; trader: string; bond: number }> {
    const seized: Array<{ hash: string; trader: string; bond: number }> = [];
    for (const [hash, entry] of this.commits) {
      if (currentBatch > entry.expireAtBatch) {
        seized.push({ hash, trader: entry.trader, bond: entry.bondLamports });
        this.seizedBonds.push({
          hash,
          trader: entry.trader,
          bond: entry.bondLamports,
          batchNum: currentBatch,
        });
        this.commits.delete(hash);
      }
    }
    return seized;
  }

  pendingCount(): number {
    return this.commits.size;
  }

  totalSeizedBonds(): number {
    let total = 0;
    for (const s of this.seizedBonds) total += s.bond;
    return total;
  }
}

File: programs/flash-book/src/matcher/commit_reveal.rs
//! Commit-reveal protocol — sequencer-proof MEV resistance.
//!
//! Phase 1 (block N):       trader submits hash(side ‖ size ‖ limit ‖ nonce ‖ trader)
//! Phase 2 (block ≤ N + K): trader reveals all parts; matcher checks hash
//! Phase 3 (next batch):    revealed order enters the FBA buffer
//!
//! Rust port uses `anchor_lang::solana_program::keccak::hashv` for the
//! commit hash — keccak is Solana-native and cheap on-chain.

use super::lot::{BaseLots, Ticks};
use super::order::{Order, OrderType, Side};
use crate::errors::FlashBookError;
use crate::state::CommitRow;
use anchor_lang::prelude::*;
use anchor_lang::solana_program::keccak::hashv;

#[derive(Debug, Clone, Copy)]
pub struct RevealPayload {
    pub trader: Pubkey,
    pub side: Side,
    pub size: BaseLots,
    pub limit: Ticks,
    pub nonce: [u8; 32],
}

impl RevealPayload {
    /// Compute the commit hash for this payload.
    pub fn hash(&self) -> [u8; 32] {
        let side_byte: [u8; 1] = [self.side as u8];
        let size_bytes = self.size.0.to_le_bytes();
        let limit_bytes = self.limit.0.to_le_bytes();
        let trader_bytes = self.trader.to_bytes();
        let h = hashv(&[
            &trader_bytes,
            &side_byte,
            &size_bytes,
            &limit_bytes,
            &self.nonce,
        ]);
        h.0
    }
}

/// Search a commit buffer for a matching active row, identified by hash.
pub fn find_commit<'a>(
    commits: &'a [CommitRow],
    hash: &[u8; 32],
) -> Option<&'a CommitRow> {
    commits.iter().find(|r| r.valid == 1 && &r.hash == hash)
}

pub fn find_commit_mut<'a>(
    commits: &'a mut [CommitRow],
    hash: &[u8; 32],
) -> Option<&'a mut CommitRow> {
    commits
        .iter_mut()
        .find(|r| r.valid == 1 && &r.hash == hash)
}

/// Insert a commit into the first empty slot. Returns Err if buffer is full.
pub fn register_commit(
    commits: &mut [CommitRow],
    hash: [u8; 32],
    trader: Pubkey,
    bond: u64,
    current_batch: u64,
    expire_in_batches: u64,
) -> Result<()> {
    if commits.iter().any(|r| r.valid == 1 && r.hash == hash) {
        return Err(error!(FlashBookError::CommitDuplicate));
    }
    for slot in commits.iter_mut() {
        if slot.valid == 0 {
            *slot = CommitRow {
                hash,
                trader,
                bond,
                committed_at_batch: current_batch,
                expire_at_batch: current_batch.saturating_add(expire_in_batches),
                valid: 1,
            };
            return Ok(());
        }
    }
    Err(error!(FlashBookError::BufferFull))
}

/// Verify a reveal against a stored commit. On success: clears the row,
/// returns a synthesized `Order`.
pub fn redeem_reveal(
    commits: &mut [CommitRow],
    payload: &RevealPayload,
    current_batch: u64,
    seq: u64,
) -> Result<Order> {
    let h = payload.hash();
    let row = find_commit_mut(commits, &h).ok_or_else(|| error!(FlashBookError::CommitMismatch))?;
    if row.trader != payload.trader {
        return Err(error!(FlashBookError::CommitMismatch));
    }
    if current_batch > row.expire_at_batch {
        return Err(error!(FlashBookError::CommitExpired));
    }
    // Clear the slot.
    *row = CommitRow::default();
    Ok(Order {
        id: seq,
        trader: payload.trader,
        side: payload.side,
        order_type: OrderType::Taker,
        size: payload.size,
        limit_price: payload.limit,
        seq,
        post_only: false,
    })
}

/// Sweep expired commits and return the seized bonds.
pub fn sweep_expired(commits: &mut [CommitRow], current_batch: u64) -> u64 {
    let mut seized: u64 = 0;
    for slot in commits.iter_mut() {
        if slot.valid == 1 && current_batch > slot.expire_at_batch {
            seized = seized.saturating_add(slot.bond);
            *slot = CommitRow::default();
        }
    }
    seized
}

pub fn pending_count(commits: &[CommitRow]) -> usize {
    commits.iter().filter(|r| r.valid == 1).count()
}
